egui_demo_lib/src/apps/demo/plot_demo.rs

```diff
@@ -309,6 +311,117 @@ impl Widget for &mut LegendDemo {
     }
 }
 
+#[derive(PartialEq, Default)]
+struct CustomAxisDemo {}
+
+impl CustomAxisDemo {
+    fn logistic_fn() -> Line {
+        let values = Values::from_explicit_callback(
+            move |x| 1.0 / (1.0 + (-2.5 * (x - 2.0)).exp()),
+            0.0..5.0,
+            100,
+        );
+        Line::new(values)
+    }
+
+    #[allow(clippy::needless_pass_by_value)]
+    fn x_grid(input: GridInput) -> Vec<GridMark> {
+        // Note: this always fills all possible marks. For optimization, `input.bounds`
+        // could be used to decide when the low-interval grids (minutes) should be added.
+
+        let mut marks = vec![];
+
+        let steps_per_day = 24 * 12;
+        let fsteps_per_day = steps_per_day as f64;
+
+        let (min, max) = input.bounds;
+        let min = (min * fsteps_per_day).floor() as i32;
+        let max = (max * fsteps_per_day).ceil() as i32;
+
+        for i in min..=max {
+            let step_size = if i % steps_per_day == 0 {
+                // 1 day
+                1.0
+            } else if i % 12 == 0 {
+                // 1 hour
+                1.0 / 24.0
+            } else {
+                // 5min
+                1.0 / fsteps_per_day
+            };
+
+            let value = i as f64 / fsteps_per_day;
+
+            marks.push(GridMark { value, step_size });
+        }
+
+        marks
+    }
+}
+
+impl Widget for &mut CustomAxisDemo {
+    fn ui(self, ui: &mut Ui) -> Response {
+        // Make sure floor() still rounds down, even in presence of arithmetic imprecision
+        const EPS: f64 = 1e-6;
```

Comment:
An alternative is to use minutes (or seconds) as the base unit (instead of days). That would make the precision problems go away since `f32` can express all integers up to 2^24 perfectly.

Reply:
Ok, I tried that approach, let me know what you think. Maybe some parts can be further simplified...